### PR TITLE
[FIX] phone_validation: extend implementation of phone_mobile_search

### PR DESCRIPTION
**Pull request by abla001**
PURPOSE

Extend implementation of ``phone_mobile_search`` that is currently limited
to performing an ilike with a term containing at least 3 characters. Any
operator is automatically considered as being an 'ilike'.

SPECIFICATIONS

Correctly support boolean check (is set / is not set), with its own small
implementation easier than the remaining computation.

Correctly propagate operator in generated SQL: like, ilike, =, those operators
should not be automatically considered as always ilike.

Correctly support negative operators. Supporting only positive search is a
bit limited.

Add tests.

Task-3012789

### CODE REVIEW

**Review by nle-odoo — COMMENTED** [review]
it seems to me that the override is meant to fix phone number when searching for the phone number (and is only meant for that)

if we want to expand it (which make sense to me, seeing the given scenario), I don't think you need to optimize that in SQL, and could just do instead:

```py
if not value and operator in ['=', '!=']:
    return [('phone', operator, value)] # or False instead of value possibly
```
side note: if we do this, we should probably also handle `[('phone_mobile_search', '=', True)]` and `[('phone_mobile_search', '!=', True)]`, so something like: `value in (False, True) and operator in ('!=', '=')`

**Comment by nle-odoo** [comment]
edit: sorry, I saw after that the original search was searching both mobile and phone. so logically if we made this work we should have the domain handle both of these field (so adding AND for negative search `(= False or != True)`, and OR for positive search `(= True or != False)`)

**Comment by abla001** [comment]
@nle-odoo totally agree,n thank you for feedback ! I updated the commit

**Review by nle-odoo — APPROVED** [review]
looks good to me

this still doesn't take into account `'=', True` and `'!=', True`. But from my test in the interface there is only `'!=', False` and `'=', False`, so it's not too bad to not have them.

**Comment by abla001** [comment]
@tde-banana-odoo can you take a look at this fix ?

File: addons/test_mail_full/tests/test_mail_thread_phone.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            5
+        )
+        cls.test_phone_records += cls.env['mail.test.sms.bl'].create([
+            {
+                'phone_nbr': '+32475110505',
+                'mobile_nbr': '+32475000505',
+            }, {
+                'phone_nbr': '0032475110606',
+                'mobile_nbr': '0032475000606',
+            }
```
Comment (std-odoo) [inline addons/test_mail_full/tests/test_mail_thread_phone.py:26]: :disguised_face: 
```suggestion
            },
```

(same for other method arguments in this file)

File: addons/test_mail_full/tests/test_mail_thread_phone.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+from odoo.tests import tagged, users
+
+
+@tagged('mail_thread')
+class TestMailThreadPhone(TestMailFullCommon, TestMailFullRecipients):
+    """ Test mail.thread.phone mixin, its tools and API """
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestMailThreadPhone, cls).setUpClass()
```
Comment (std-odoo) [inline addons/test_mail_full/tests/test_mail_thread_phone.py:14]:
```suggestion
        super().setUpClass()
```

File: addons/test_mail_full/tests/test_mail_thread_phone.py [hunk 3]
```diff
@@ -36,3 +36,32 @@ def test_initial_data(self):
...
+
+        # test "=" search
+        results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', '=', '0475')])
+        results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', '=', '+32475')])
+        self.assertEqual(results, test_phone_records[5:])
+
+        # test partial not like search
+        results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', 'not ilike', '0475')])
+        results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', 'not ilike', '101')])
+        results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', 'not ilike', '606')])
```
Comment (std-odoo) [inline addons/test_mail_full/tests/test_mail_thread_phone.py:67]: I think you forgot to check `results` here

File: addons/phone_validation/models/mail_thread_phone.py [hunk 4]
```diff
@@ -47,14 +47,20 @@ class PhoneMixin(models.AbstractModel):
...
         ]
         if not phone_fields:
             raise UserError(_('Missing definition of phone fields.'))
 
-        value = value.strip()
+        # search if phone/mobile is set or not
+        if (value is True or not value) and operator in ['=', '!=']:
+            final_operator = '=' if (operator == '=' or value) else '!='
+            op = '&' if final_operator == '=' else '|'
+            return [op] * (len(phone_fields) - 1) + [(phone_field, final_operator, False) for phone_field in phone_fields]
```
Comment (std-odoo) [inline addons/phone_validation/models/mail_thread_phone.py:62]: a bit more readable with expression.AND / expression.OR imo
```suggestion
        # search if phone/mobile is set or not
        if (value is True or not value) and operator in ('=', '!='):
            if value:
                # inverse the operator
                operator = '=' if operator == '!=' else '!='
            op = expression.AND if operator == '=' else expression.OR
            return op([[(phone_field, operator, False)] for phone_field in phone_fields])
```

(and annoying std, tuple better than list for this case)

File: addons/test_mail_full/tests/test_mail_thread_phone.py [hunk 5]
```diff
@@ -65,3 +65,18 @@ def test_search_phone_mobile_search(self):
...
+        # test Falsy -> is set / is not set
+        for test_values in [False, '', ' ']:
+            results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', '=', test_values)])
+            self.assertFalse(results)
+            results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', '!=', test_values)])
+            self.assertEqual(results, test_phone_records)
+
+        # is not set -> != True
+        results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', '!=', True)])
+        self.assertFalse(results)
```
Comment (std-odoo) [inline addons/test_mail_full/tests/test_mail_thread_phone.py:82]: with my proposition above, the "opposite" will work `(..., '=', True)` (currently it return same result as `... == False`)

could be nice to add a test

File: addons/test_mail_full/tests/test_mail_thread_phone.py [hunk 6]
```diff
@@ -65,3 +65,18 @@ def test_search_phone_mobile_search(self):
         results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', 'not ilike', '0475')])
         results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', 'not ilike', '101')])
         results = self.env['mail.test.sms.bl'].search([('phone_mobile_search', 'not ilike', '606')])
+
+    @users('employee')
+    def test_search_phone_mobile_search_boolean(self):
```
Comment (std-odoo) [inline addons/test_mail_full/tests/test_mail_thread_phone.py:70]: the test doesn't check for the "and" or the "or" operator 
this line:
```
op = '&' if final_operator == '=' else '|'
```

(could be nice to set `mobile_nbr` to False on one of the records)

File: addons/phone_validation/models/mail_thread_phone.py [hunk 7]
```diff
@@ -55,38 +62,35 @@ def _search_phone_mobile_search(self, operator, value):
...
-                    REGEXP_REPLACE(model.mobile, %s, '', 'g') ILIKE %s;
+                WHERE {where_str};
             """
             term = '%' + re.sub(pattern, '', value) + '%'
-            self._cr.execute(query, (
-                pattern, term,
-                pattern, term
+            self._cr.execute(query, tuple(
+                [pattern, term] * len(phone_fields)
             ))
```
Comment (std-odoo) [inline addons/phone_validation/models/mail_thread_phone.py:100]: could be onelined while you change this line

**Review by std-odoo — COMMENTED** [review]
@abla001 @tde-banana-odoo 

Hi :) 

Nice fix, I have 2-3 propositions :)

**Review by tde-banana-odoo — COMMENTED** [review]
Hello @std-odoo ,

Thanks for the suggestions, code updated :) while being at it I also updated the code to correctly support negative operators (was always forced to ilike before) as well as direct check ('=' was always considered as 'like').

Cheers !

File: addons/phone_validation/models/mail_thread_phone.py [hunk 8]
```diff
@@ -55,39 +63,34 @@ def _search_phone_mobile_search(self, operator, value):
...
-                    REGEXP_REPLACE(model.phone, %s, '', 'g') ILIKE %s OR
-                    REGEXP_REPLACE(model.mobile, %s, '', 'g') ILIKE %s;
+                WHERE {where_str};
             """
             term = '%' + re.sub(pattern, '', value) + '%'
-            self._cr.execute(query, (
-                pattern, term,
-                pattern, term
-            ))
+            self._cr.execute(query, tuple([pattern, term] * len(phone_fields)))
```
Comment (std-odoo) [inline addons/phone_validation/models/mail_thread_phone.py:117]:
```suggestion
            self._cr.execute(query, (pattern, term) * len(phone_fields)))
```

File: addons/phone_validation/models/mail_thread_phone.py [hunk 9]
```diff
@@ -55,39 +63,34 @@ def _search_phone_mobile_search(self, operator, value):
...
             """
             term = re.sub(pattern, '', value[1 if value.startswith('+') else 2:]) + '%'
-            self._cr.execute(query, (
-                pattern, '00' + term,
-                pattern, '+' + term,
-                pattern, '00' + term,
-                pattern, '+' + term
+            self._cr.execute(query, tuple(
+                [pattern, '00' + term, pattern, '+' + term] * len(phone_fields)
             ))
```
Comment (std-odoo) [inline addons/phone_validation/models/mail_thread_phone.py]:
```suggestion
            self._cr.execute(query, (
                pattern, '00' + term, pattern, '+' + term * 
            ) * len(phone_fields))
```

File: addons/phone_validation/models/mail_thread_phone.py [hunk 10]
```diff
@@ -68,36 +68,52 @@ def _search_phone_mobile_search(self, operator, value):
...
+                    )"""
+                    for phone_field in phone_fields
+                )
+            query = f"SELECT model.id FROM {self._table} model WHERE {where_str};"
+
+            term = re.sub(pattern, '', value[1 if value.startswith('+') else 2:])
+            if operator not in ('=', '!='):  # for like operators
+                term = f'{term}%'
+            self._cr.execute(
+                query, tuple([pattern, '00' + term, pattern, '+' + term] * len(phone_fields))
```
Comment (std-odoo) [inline addons/phone_validation/models/mail_thread_phone.py:100]:
```suggestion
                query, (pattern, '00' + term, pattern, '+' + term) * len(phone_fields)
```

**Review by std-odoo — COMMENTED** [review]
@tde-banana-odoo Some last polishing comments, otherwise LGTM :)

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Pull request closed without merging**